### PR TITLE
Add toggle switch component

### LINKED ISSUE #519: Switch/toggle
Vi jobber med å få ut Push notifikasjoner i mobilbanken, og har behov for en komponent som har lik design og oppførsel som en native switch/toggle. 

I første skisse så har vi brukt en checkmark men det blir litt misvisende for konteksten vår. 

checkbox

Forslag til ny komponent i bruk:

push-switch

### PR DESCRIPTION
**Pull request by antidecaf**
This PR adds a new toggle switch component. It looks like this:

Default:
image

Checked/selected:
image

**It's not yet decided whether or not this will be merged.** Some questions have arisen along the way and need to be answered first:

- Does the toggle switch pattern make sense in a desktop UI? This component renders a switch on any screen size. Should the switch in stead be replaced by a different pattern on large screens, and if so, which one?

- Guidelines are not yet written, but should be if this is to be merged. When to use a toggle switch vs radio buttons/checkboxes should be very clear. According to [NN Group guidelines](https://www.nngroup.com/articles/toggle-switch-guidelines/) the toggle switch pattern demands immediate results, i.e. they should not be used in forms that require users to click a button for changes to take effect. Any input on this? Anything else?

- How to handle the focus state? This implementation includes the regular focus outline (same one used in form elements), but it feels a bit off given that switches generally don't have an outline in a native mobile context. Do we:

1. Keep it as is?
2. Show the outline only on large screens?
3. Remove the outline altogether? (If so, what consequenses does it have for accessibility?)

- How about the hover state? I've added the focus style on hover for now.

Hover/focus state:
image
image

Any input is welcome!

Fixes #519

### CODE REVIEW

**Comment by ivarni** [comment]
I don't think focus-states are very relevant on mobile? I'm not even sure how I focus an input control without triggering it on my phone. Or are there accessibility tools out there that rely on focus styling?

File: packages/ffe-form-react/src/ToggleSwitch.js
Comment (selbekk) [inline packages/ffe-form-react/src/ToggleSwitch.js]: This will receive a new ID on each render. Instead, use a class component, and set the ID as an instance variable

File: packages/ffe-form-react/src/ToggleSwitch.js
Comment (selbekk) [inline packages/ffe-form-react/src/ToggleSwitch.js]: perhaps `alignRight` is a more descriptive name?

Comment (selbekk) [reply packages/ffe-form-react/src/ToggleSwitch.js]: Also - `id` is missing from the propTypes.

Comment (selbekk) [reply packages/ffe-form-react/src/ToggleSwitch.js]: Also also - should `value` also be specified here? It's the main API to the component, could be nice to include in the docs?

File: packages/ffe-form-react/src/ToggleSwitch.md
Comment (selbekk) [inline packages/ffe-form-react/src/ToggleSwitch.md:21]: :joy:

**Review by selbekk — CHANGES_REQUESTED** [review]
All in all, this looks very good! Should we also add some usage information, like when this is a good choice over the very similar RadioSwitch?

**Comment by antidecaf** [comment]
> I don't think focus-states are very relevant on mobile? I'm not even sure how I focus an input control without triggering it on my phone. Or are there accessibility tools out there that rely on focus styling?

You're right, it's mainly an issue in desktop browsers. But on a second glance it looks less weird than anticipated with the hover state that is identical to the focus state.

File: packages/ffe-form-react/src/ToggleSwitch.js
Comment (antidecaf) [reply packages/ffe-form-react/src/ToggleSwitch.js]: Will fix!

File: packages/ffe-form-react/src/ToggleSwitch.js
Comment (antidecaf) [reply packages/ffe-form-react/src/ToggleSwitch.js]: Will fix!

**Comment by antidecaf** [comment]
> Should we also add some usage information, like when this is a good choice over the very similar RadioSwitch?

Yes, as mentioned in the original comment, guidelines need to be added before this is merged. I'm not sure what exactly to add however, so I'll await some feedback from UX before doing any of that. :)

**Review by kwltrs — DISMISSED** [review]

**Review by hodoea — CHANGES_REQUESTED** [review]
skjermbilde 2018-12-07 kl 13 55 28


Kan dette fungere som guidelines?

**Comment by selbekk** [comment]
Veldig bra @hodoea ! Jeg savner fortsatt en differensiering mellom når man skal bruke en toggle vs en radio switch, though - kanskje verdt å nevne eksplisitt?

**Comment by hodoea** [comment]
My two cents on these questions:

>It's not yet decided whether or not this will be merged. Some questions have arisen along the way and need to be answered first:

>Does the toggle switch pattern make sense in a desktop UI? This component renders a switch on any screen size. Should the switch in stead be replaced by a different pattern on large screens, and if so, which one?

I am bit unsure here, I dont see any problem with using this for desktop as long as we can convey to people how to best use this in the guidelines. But many of my fellow interaction designers don`t agree. They propose that we look for something else on desktop. These are some  examples from Google forms where you use a toggle-switch to "turn on" more features in your form on desktop. And the other example is where you use it to state that something is required, I dont understand why they use this here though...

However I`ve mostly seen the use of toggle-switches for native functionalities (settings on your phone++)

screen recording 2018-12-06 at 12 59 pm

And another example:
screen recording 2018-12-06 at 01 00 pm

>Guidelines are not yet written, but should be if this is to be merged. When to use a toggle switch vs radio buttons/checkboxes should be very clear. According to NN Group guidelines the toggle switch pattern demands immediate results, i.e. they should not be used in forms that require users to click a button for changes to take effect. Any input on this? Anything else?

>How to handle the focus state? This implementation includes the regular focus outline (same one used in form elements), but it feels a bit off given that switches generally don't have an outline in a native mobile context. Do we:

>Keep it as is?
>Show the outline only on large screens?

I propose we show the outline on large screens only if that is possible

>Remove the outline altogether? (If so, what consequenses does it have for accessibility?)

>How about the hover state? I've added the focus style on hover for now.

On mobile you do not have a hover state. I think perhaps this is one of the challenges if we use this component for both mobile and desktop.

**Comment by antidecaf** [comment]
Yes, the hover/focus states are only present on large screens. Looks fine to me now:

ezgif-1-ea1f06cdb031

I've added the guidelines, but won't merge the PR for now, as it has been decided that this component needs to be tested in an app first.

**Comment by hodoea** [comment]
@selbekk hm.. I have not used this component before, and I read what we have of guidelines for this. I dont see why we should use this over a radio button in our forms. I have raised this in our fora on Slack. One designer mentioned that they have many questions in their forms where they have used radio switches. Because of the components size the questions are "drowning" in the form. 
Maybe we can fix this by adding to the guidelines that if you have more than x questions in a form, consider using radio buttons?

Where is a radio switch a very good use? Maybe anyone knows where the need for this come from.. ?

**Comment by antidecaf** [comment]
The RadioSwitch component differs from ToggleSwitch in that it has no default value, and is used as a part of longer forms with submit buttons. Apart from that I figure it has an entirely different feel. ToggleSwitches are designed to imitate physical switches, which I'm assuming is why the strict on/off usage guidelines make sense.

**Comment by hodoea** [comment]
Yes ofc forgot the default value. I was just a little bit confused because of the second statement in the guidelines: typisk i formen "ja eller nei", **eller "av eller på"**.

**Comment by hodoea** [comment]
Yes ofc, I was just a little bit confused because of the second statement in the guidelines: typisk i formen "ja eller nei", **eller "av eller på"*.

**Comment by hodoea** [comment]
Nvm, Adi shared this on the background for it: 

_Radio switch var laget til å tydeligjøre lett valg på mobil. Det var laget for ca. 4 år siden som en del av NN (nye nettsider) og DNA (Forsikring) prosjekter akkurat i starten av å jobbe med den nye profil refresh. Det var testet i bilforsikring kjøpsløp på mobil og fungerte skikkelig bra pga at det var veldig tydelig valg og veldig oversiktlig, plus at det var veldig mobilvennlig. Det var mange runder ift utforming av knappen og det som ble til slutt var en blanding av en som eksisterte før og ny stilen. Jeg synes det er ikke den fineste komponent i at det er for tungt i mørk blå men det kunne vi justere. Ideen er å bare bruke det til lett valg, som “Ja/Nei”. Så snart som det er mer en 3 valg bør man bruke vanlig radio knapper (i en list nedover). Det har alltid vært et spørsmål om hvis vi trenger å ha en så stor og tydelig variant ift helhetlig visuell hierarkiet (f.eks i dag konkurrere den mørke blå med alle andre action knapper på sidene) så det kunne man godt se på. Grunnen at det er mørkt blå er at ideen var at alle “valgt” states burde være likt på tvers av designet og akkurat da brukte vi mørk blå mye mer i designet vårt. I dagens design er det litt for tungt i sammenheng med alle andre knapper osv og kunne være f.eks en mer subtil lys blå, eller bare bruke vanlig radio knapper men passe på at de har en stor klikkflate._
 

I can just add in the documentation for radio switch that if there are more than 3 choices in your list, use radio buttons instead. This was sort of a sidetrack, sorry 😅

**Comment by selbekk** [comment]
Perhaps the difference can be explained like this?

The toggle switch is for "on / off" choices, typically found in settings, while the radio switch is for `yes / no" choices, typically found in forms with a submit button?

**Comment by hodoea** [comment]
Takk. Har allerede skrevet i første avsnitt at man ikke skal bruke toggle-switch dersom man har bruk for en send/lagre knapp i tillegg. Men jeg la det til igjen i forslaget ditt uansett. @selbekk  


skjermbilde 2018-12-07 kl 12 55 29

Sorry for at teksten var i et bilde i starten, har kopiert inn teksten nå :) @antidecaf 

Bruk av toggle-switch

Toggle-switcher er av/på switcher. De tilbyr brukerne valget mellom to avhengige valg og de har alltid en default verdi. De skal brukes i kontekster hvor systemet gir umiddelbare resultater. Det skal ikke brukes dersom du har behov for en lagre/send knapp i tillegg. Navnet du bruker for å beskrive handlingen skal være kort og konsist. 

Toggle-switcher kan erstatte to radioknapper eller én enkelt checkbox når en bruker skal velge mellom to avhengige states.  

Toggle-switch vs radio switch
Bruk en toggle-switch for «av/på» valg, dette er for eksempel vanlig i bl.a innstillinger. Radio switcher derimot brukes for «Ja/Nei» valg som bl.a finnes i skjemaer hvor du i tillegg har behov for en lagre/send knapp.

**Comment by selbekk** [comment]
som jeg skulle gjort det selv @hodoea :laughing: tusen takk!

**Comment by ivarni** [comment]
@antidecaf Are this one pretty much ready to go once the documentation is added?

**Comment by antidecaf** [comment]
@ivarni Yeah, pretty much. But it doesn't have a green light from UX at the moment, it's still being tested in one app afaik. Not sure if it will be merged or not in the end, but for now we can leave it.

**Pull request closed without merging**
